// bitgo/dist/src/v2/coins/susd.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var BaseCoin = require('../baseCoin');
var Susd = /** @class */ (function (_super) {
    __extends(Susd, _super);
    function Susd() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Returns the factor between the base unit and its smallest subdivison
     * @return {number}
     */
    Susd.prototype.getBaseFactor = function () {
        return 1e2;
    };
    Susd.prototype.getChain = function () {
        return 'susd';
    };
    Susd.prototype.getFamily = function () {
        return 'susd';
    };
    Susd.prototype.getFullName = function () {
        return 'Silvergate USD';
    };
    /**
     * Return whether the given m of n wallet signers/ key amounts are valid for the coin
     */
    Susd.prototype.isValidMofNSetup = function (_a) {
        var m = _a.m, n = _a.n;
        return m === 0 && n === 0;
    };
    return Susd;
}(BaseCoin));
module.exports = Susd;
