// bitgo/dist/src/v2/coins/tbtg.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var Btg = require('./btg');
var Tbtg = /** @class */ (function (_super) {
    __extends(Tbtg, _super);
    function Tbtg() {
        // TODO: move to bitgo-utxo-lib (BG-6821)
        return _super.call(this, {
            messagePrefix: '\x18Bitcoin Signed Message:\n',
            bech32: 'tb',
            bip32: {
                public: 0x043587cf,
                private: 0x04358394
            },
            pubKeyHash: 0x6f,
            scriptHash: 0xc4,
            wif: 0xef,
            coin: 'btg'
        }) || this;
    }
    Tbtg.prototype.getChain = function () {
        return 'tbtg';
    };
    Tbtg.prototype.getFullName = function () {
        return 'Testnet Bitcoin Gold';
    };
    return Tbtg;
}(Btg));
module.exports = Tbtg;
