// bitgo/dist/src/v2/coins/tdash.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var Dash = require('./dash');
var Tdash = /** @class */ (function (_super) {
    __extends(Tdash, _super);
    function Tdash() {
        // TODO: move to bitgo-utxo-lib (BG-6821)
        return _super.call(this, {
            messagePrefix: '\x19Dash Signed Message:\n',
            bip32: {
                public: 0x0488b21e,
                private: 0x0488ade4
            },
            pubKeyHash: 0x8c,
            scriptHash: 0x13,
            wif: 0xef,
            dustThreshold: 0,
            dustSoftThreshold: 100000,
            feePerKb: 100000,
            coin: 'dash'
        }) || this;
    }
    Tdash.prototype.getChain = function () {
        return 'tdash';
    };
    Tdash.prototype.getFullName = function () {
        return 'Testnet Dash';
    };
    return Tdash;
}(Dash));
module.exports = Tdash;
